src/algebra/ring/basic.lean

```diff
@@ -825,6 +826,35 @@ by rw [← mul_self_eq_mul_self_iff, one_mul]
 lemma units.inv_eq_self_iff (u : units α) : u⁻¹ = u ↔ u = 1 ∨ u = -1 :=
 by { rw inv_eq_iff_mul_eq_one, simp only [units.ext_iff], push_cast, exact mul_self_eq_one_iff }
 
+/-- Makes a ring homomorphism from a additive group homomorphism of a commutative rings and an
+    integral domain that commutes with self multiplication. -/
+def ring_hom.mk_mul_self_of_two_ne_zero [comm_ring β] (f : β →+ α) (h : ∀ x, f (x * x) = f x * f x)
+  (h2 : (2 : α) ≠ 0) (h_one : f 1 = 1) : β →+* α :=
+{ map_one' := h_one,
+  map_mul' := begin
+    intros x y,
+    have hxy := h (x + y),
+    simp only [mul_add, add_mul, h x, h y, f.map_add] at hxy,
+    rw [← sub_eq_zero_iff_eq] at hxy,
+    rw [add_comm, ← sub_sub, ← sub_sub, ← sub_sub, mul_comm y x, mul_comm (f y) (f x)] at hxy,
+    simp [add_assoc, add_sub_assoc] at hxy,
+    rw [sub_sub, ← two_mul, ← add_sub_assoc, ← two_mul, ← mul_sub,
+      mul_eq_zero, sub_eq_zero_iff_eq, or_iff_not_imp_left] at hxy,
+    exact hxy h2,
+  end,
+  ..f }
+
+@[simp]
+lemma ring_hom.mk_mul_self_of_two_ne_zero_to_fun [comm_ring β] (f : β →+ α)
+  (h : ∀ x, f (x * x) = f x * f x) (h2 : (2 : α) ≠ 0) (h_one : f 1 = 1) :
+  (ring_hom.mk_mul_self_of_two_ne_zero f h h2 h_one).to_fun = f :=
```

Comment:
Isn't the simp-normal form
```suggestion
  ring_hom.mk_mul_self_of_two_ne_zero f h h2 h_one = f :=
```
I'd expect the linter to complain about this spelling

Reply:
No complaints from the linter, I did change it to a nicer looking version though.